iml-manager-cli/src/server.rs

```diff
@@ -218,250 +219,400 @@ async fn wait_till_agent_starts(
-                hosts.objects.into_iter().map(|h| {
-                    vec![
-                        h.id.to_string(),
-                        h.fqdn,
-                        h.state,
-                        h.nids.unwrap_or_else(|| vec![]).join(" "),
-                    ]
-                }),
+                &["Description", "Test", "Error"],
+                failed_tests
+                    .into_iter()
+                    .map(|x| vec![&x.description, &x.test, &x.error]),
             );
 
             table.printstd();
+
+            term.write_line("\n\n")?;
         }
-        ServerCommand::Add(config) => {
-            let term = Term::stdout();
 
-            let new_hosts = hostlist_parser::parse(&config.hosts)?;
+        if !can_continue(&config) {
+            return Err(ImlManagerCliError::ApiError(
+                "Did not deploy profile.".into(),
+            ));
+        }
+    }
 
-            tracing::debug!("Parsed hosts {:?}", new_hosts);
+    Ok(())
+}
 
-            let (profiles, api_hosts): (ApiList<ServerProfile>, ApiList<Host>) =
-                future::try_join(get_all(), get_all()).await?;
+fn get_host_profile_configs<'a>(
+    profile: &'a ServerProfile,
+    objects: &[(u32, Vec<ProfileTest>)],
+) -> Vec<HostProfileConfig<'a>> {
+    objects
+        .iter()
+        .map(|(host, _)| HostProfileConfig {
+            host: *host,
+            profile: &profile.name,
+        })
+        .collect()
+}
 
-            let profile_opt = profiles
-                .objects
-                .into_iter()
-                .filter(|x| x.user_selectable)
-                .find(|x| x.name == config.profile);
+async fn post_profile<'a>(
+    objects: Vec<HostProfileConfig<'a>>,
+) -> Result<Objects<HostProfileCmdWrapper>, ImlManagerCliError> {
+    Ok(post(HostProfile::endpoint_name(), Objects { objects })
+        .await?
+        .json()
+        .await
+        .map_err(iml_manager_client::ImlManagerClientError::Reqwest)?)
+}
 
-            let profile = match profile_opt {
-                Some(p) => p,
-                None => {
-                    display_error(format!("Unknown profile {}", config.profile));
+fn get_commands_from_wrapper(objects: Vec<HostProfileCmdWrapper>) -> Vec<Command> {
+    objects
+        .into_iter()
+        .flat_map(|HostProfileCmdWrapper { commands }| commands)
+        .collect()
+}
 
-                    return Ok(());
-                }
-            };
+async fn profiles_and_servers(
+    config: &AddHosts,
+    new_hosts: BTreeSet<String>,
+) -> Result<(ServerProfile, BTreeSet<String>), ImlManagerCliError> {
+    let (profiles, api_hosts): (ApiList<ServerProfile>, ApiList<Host>) =
+        future::try_join(get_all(), get_all()).await?;
+
+    let profile_name = config.profile.clone();
+    let profile = get_profile(&config, profiles).ok_or_else(|| {
+        Error::new(
+            ErrorKind::NotFound,
+            format!("{} profile not found.", profile_name),
```

Comment:
Here you handle `None` and duplicate the same message as in `get_profile`.

Reply:
While the messages are basically the same, I believe they are mutually exclusive. The difference is that the first message is meant to display on the cli while the NotFound error message is attached to the NotFound error. With that said, if the error case is hit and it automatically takes care of calling the display method to display the error message then I can probably get rid of the conditional statement that prints the message.